**Note to grader:** Each question consists of parts, e.g. Q1(i), Q1(ii), etc. Each part must be graded  on a 0-4 scale, following the standard NJIT convention (A:4, B+: 3.5, B:3, C+: 2.5, C: 2, D:1, F:0). 
The total score must be re-scaled to 100 &mdash; that should apply to all future assignments so that Canvas assigns the same weight on all assignments. 



In [1]:
# Grader's area
import numpy as np
M = np.zeros([10,10])
maxScore = 0

# **Assignment 4**

The goal of this assignment is to run some experiments with scikit-learn on a fairly sizeable and interesting image data set. This is the MNIST data set, which consists of tens of thousands of images, each having 28x28 pixels. By today's standards, this dataset may seem relatively tiny, but only a few years ago was quite challenging computationally, and it motivated the development of several ML algorithms and models that are now state-of-the-art  solutions for much bigger data sets. 

The assignment is experimental. We will try to whether a combination of PCA and kNN can yield any good results for the MNIST data set. Let's see if it can be made to work on this data set. 

Note: There are less difficult Python parts in this assignment. You can get things done by just using code snippets from the class notebooks (no attribution is needed for this). But your participation and interaction via Canvas is always appreciated!

## Preparation Steps

In [2]:
# Import all necessary python packages

import numpy as np
from sklearn.datasets import fetch_openml
from sklearn.model_selection import train_test_split
from sklearn.decomposition import PCA
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score

In [3]:
# We load the data set directly from scikit learn.
# Note: this operation may take a few seconds.
# If for any reason it fails we can revert back to loading from local storage. 

from sklearn.datasets import fetch_openml
from sklearn.model_selection import train_test_split

X, y = fetch_openml('mnist_784', version=1, return_X_y=True)
y = y.astype(int)
X = ((X / 255.) - .5) * 2    # Line A for Question 1(v)
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=10000, random_state=123, stratify=y)

/usr/local/lib/python3.9/dist-packages/sklearn/datasets/_openml.py:968: FutureWarning: The default value of `parser` will change from `'liac-arff'` to `'auto'` in 1.4. You can set `parser='auto'` to silence this warning. Therefore, an `ImportError` will be raised from 1.4 if the dataset is dense and pandas is not installed. Note that the pandas parser may return different data types. See the Notes Section in fetch_openml's API doc for details.
  warn(


## <font color = 'blue'> Question 1. Inspecting the Dataset </font>

**(i)** How many data points are in the training and test sets? <br>
**(ii)** How many attributes does the data set have?

Explain how you found the answer to the first two questions. 

[**Hint**: Use the 'shape' method associated with numpy arrays.]

**(iii)** How many different labels does this data set have? Can you demonstrate how to read that number from the vector of labels *y_train*?  <br>
**(iv)** How does the number of attributes relate to the size of the images? <br>
**(v)** What is the role of Line A in the above code? 





*(Please insert cells below for your answers. Clearly identify the part of the question you are answering)*

In [4]:
### (i) How many data points are in the training and test sets?

print("y_train shape: ", y_train.shape)
print("y_test shape: ", y_test.shape)

y_train shape:  (60000,)
y_test shape:  (10000,)


There are 60,000 data points in the traning set and 10,000 data points in the test set.

In [5]:
### (ii) How many attributes does the data set have?

print("X_train shape: ", X_train.shape)
print("X_test shape: ", X_test.shape)

X_train shape:  (60000, 784)
X_test shape:  (10000, 784)


There are 784 attributes in the dataset.
<br>
The dimensions of the array are represented in the form of a tuple when we associate the shape() method with the NumPy array.

In [6]:
### (iii) How many different labels does this data set have? Can you demonstrate how to read that number from the vector of labels y_train?

Q1_iii = y_train.unique()
print('Different labels: ', Q1_iii)
print('There are', len(Q1_iii), 'different labels.')

Different labels:  [4 8 7 0 5 2 6 9 3 1]
There are 10 different labels.


There are 10 different labels in the dataset.

In [7]:
### (iv) How does the number of attributes relate to the size of the images?

Based on the assignment description, this is the MNIST data set, which consists of tens of thousands of images, each having 28x28 pixels, which means there are 784 pixels for each image.
<br>
There are 784 attributes in the dataset, therefore, each pixel information for each image is available as an attribute. This is how the number of attributes relate to the size of the images.

In [8]:
### (v) What is the role of Line A in the above code?

Since the dataset consists of tens of thousands of images, the color spectrum is represented in discrete values, where basic colors are represented in 8 bits, so they are each limited to 256 (2^8), that's the reason for dividing by 256 (255 in this case since 0 is included) before normalization.
<br>
Therefore, the role of Line A in the above code is to standarize the values of each pixel by calculating the difference from the midpoint (0.5) and dividing by 2, so the range would be between 0 and 1.

In [9]:
# For grader use only

# In this case, make excetion and, assign 0-2 points for each subquestion

# Insert grade here  
# G[1,1] = 
# G[1,2] =
# G[1,3] = 
# G[1,4] = 
# G[1,5] =  

maxScore = maxScore + 10

##  <font color = 'blue'> Question 2. PCA on MNIST </font>

Because the number of attributes of the MNIST data set may be too big to apply kNN on it (due to the 'curse of dimensionality'), we would like to compress the images down to a smaller number of transformed attributes. 

Use scikit-learn to output a data set *X_train_transformed* and *X_test_transformed*, with $l$ attributes. Here a reasonable choice of $l$ is 10, equal to the number of labels. But you can try slightly smaller or larger values as well. 


**Hint**: Take a look at [this notebook](https://colab.research.google.com/drive/1DG5PjWejo8F7AhozHxj8329SuMtXZ874?usp=drive_fs) we used in the lecture, and imitate what we did there. Be sure to use only the scikit-learn demonstration, not the exhaustive PCA steps we did before it.

**Note**: This computation can take a while! If problems are encountered, we can try the same experiment on a downsized data set. 

In [10]:
from sklearn.decomposition import PCA

Q2 = PCA(n_components = 10).fit(X_train)
X_train_transformed = Q2.transform(X_train)
X_test_transformed = Q2.transform(X_test)
print('X_train_transformed: ', X_train_transformed)
print('X_test_transformed: ', X_test_transformed)

X_train_transformed:  [[ 0.0343793   5.93042881 -1.32557812 ...  0.80217255  0.79794175
   0.24868201]
 [ 1.71107123  0.61959336 -5.61616294 ... -1.9073887   4.23969742
   0.72691489]
 [-3.70015373  6.08597842 -2.92447796 ...  3.91635367 -3.01700488
   1.98420708]
 ...
 [-1.38016639 -7.54467965  0.50226135 ...  0.45954821  0.99674125
   0.98841315]
 [ 1.20194877  7.38339809 -1.3696067  ... -0.36823836  4.36173203
   2.0124995 ]
 [-1.08511146 -4.05124447 -6.87142278 ...  5.18843696 -1.13764208
  -0.03940527]]
X_test_transformed:  [[-3.03519079e+00 -1.98450509e+00  1.13102089e+00 ... -8.83340827e-02
   3.01250125e+00 -1.37865055e+00]
 [-2.40049599e+00  6.22497731e+00  3.36797431e-03 ... -9.45511471e-01
  -1.76314790e+00  1.41808662e+00]
 [ 3.00133263e+00  1.35338455e+00 -1.65932928e-01 ...  2.03890463e+00
  -2.93489497e-01 -7.54813961e-01]
 ...
 [ 5.16858174e+00 -3.04103502e+00  2.89070528e+00 ...  5.93433907e+00
  -4.42940692e+00 -1.31594500e+00]
 [-5.94360768e+00 -1.19968886e+00 -1.181

In [11]:
# For grader use

maxScore = maxScore +4 

# insert grade here (out of 4)
# G[2,1] =

## <font color = 'blue'> Question 3. kNN on MNIST attributes from PCA </font>


Having created a *transformed* MNIST data set, we can now apply a kNN approach. Here are the steps:

(i) Fit a $k$-NN classifier on the transformed data set. Here $k$ is a hyperparameter, and you can experiment with it. Be aware though, that larger values of $k$ can take more time to fit. 

(ii) Apply the classifier on the transformed test set. What is the classification accuracy? 

(iii) A theoretical question: if we skipped all the above steps and we just assigned a **random** label to each test point, what would the classification accuracy be on average?  Is your result from (ii) better than the random expectation? 

(iv) Experiment with different settings of $k$, and other hyperparameters that are described in the scikit-learn manual of the kNN classifier. Report your findings in a separate cell. Also for **participation points**: report your best result on Canvas! 

[**Hint**: Imitate the steps from the classroom notebook]


In [12]:
### (i) Fit a  k-NN classifier on the transformed data set.

from sklearn.neighbors import KNeighborsClassifier
Q3 = KNeighborsClassifier(n_neighbors=5,
                           weights='distance',
                           algorithm='kd_tree'
                          )
Q3.fit(X_train_transformed, y_train)

KNeighborsClassifier(algorithm='kd_tree', weights='distance')

In [13]:
### (ii) Apply the classifier on the transformed test set. What is the classification accuracy?

Q3_Predicted = Q3.predict(X_test_transformed)
print('The classification accuracy is', Q3.score(X_test_transformed,y_test))

The classification accuracy is 0.9354


The classification accuracy is 0.9354 (93.54%).

In [14]:
### (iii) If we skipped all the above steps and we just assigned a random label to each test point, what would the classification accuracy be on average?
### Is your result from (ii) better than the random expectation?

from sklearn.metrics import accuracy_score
Q3_iii = np.random.choice(y.unique(),size=len(y_test))
print('If a random label is assigned to each test point, the classification accuracy is', accuracy_score(Q3_iii,y_test))

If a random label is assigned to each test point, the classification accuracy is 0.1039


If a random label is assigned to each test point, the classification accuracy is 0.1039 (10.39%) on average.
<br>
Therefore, the accuracy with the random lables classifier is much lower than the PCA followed by KNN classifier (93.54%).

In [15]:
### (iv) Experiment with different settings of  k, and other hyperparameters that are described in the scikit-learn manual of the kNN classifier.

from sklearn.neighbors import KNeighborsClassifier

knn = KNeighborsClassifier(n_neighbors=5, 
                           weights='distance',
                           algorithm='kd_tree'
                          )
knn.fit(X_train_transformed, y_train)
Q3_1 = knn.predict(X_test_transformed)

knn = KNeighborsClassifier(n_neighbors=5, 
                           weights='distance',
                           algorithm='ball_tree'
                          )
knn.fit(X_train_transformed, y_train)
Q3_2 = knn.predict(X_test_transformed)

knn = KNeighborsClassifier(n_neighbors=5,
                           weights='distance',
                            algorithm='brute'
                          )
knn.fit(X_train_transformed, y_train)
Q3_3 = knn.predict(X_test_transformed)

knn = KNeighborsClassifier(n_neighbors=5,
                           weights='uniform',
                           algorithm='kd_tree'
                          )
knn.fit(X_train_transformed, y_train)
Q3_4 = knn.predict(X_test_transformed)

knn = KNeighborsClassifier(n_neighbors=5,
                           weights='uniform',
                           algorithm='ball_tree'
                          )
knn.fit(X_train_transformed, y_train)
Q3_5 = knn.predict(X_test_transformed)

knn = KNeighborsClassifier(n_neighbors=5,
                           weights='uniform',
                            algorithm='brute'
                          )
knn.fit(X_train_transformed, y_train)
Q3_6 = knn.predict(X_test_transformed)

print('Accuracy for k = 5, weights = distance, algorithm = kd_tree', accuracy_score(y_test, Q3_1))
print('Accuracy for k = 5, weights = distance, algorithm = ball_tree', accuracy_score(y_test, Q3_2))
print('Accuracy for k = 5, weights = distance, algorithm = brute', accuracy_score(y_test, Q3_3))
print('Accuracy for k = 5, weights = uniform, algorithm = kd_tree', accuracy_score(y_test, Q3_4))
print('Accuracy for k = 5, weights = uniform, algorithm = ball_tree', accuracy_score(y_test, Q3_5))
print('Accuracy for k = 5, weights = uniform, algorithm = brute', accuracy_score(y_test, Q3_6))

Accuracy for k = 5, weights = distance, algorithm = kd_tree 0.9354
Accuracy for k = 5, weights = distance, algorithm = ball_tree 0.9354
Accuracy for k = 5, weights = distance, algorithm = brute 0.9354
Accuracy for k = 5, weights = uniform, algorithm = kd_tree 0.9347
Accuracy for k = 5, weights = uniform, algorithm = ball_tree 0.9347
Accuracy for k = 5, weights = uniform, algorithm = brute 0.9347


In [16]:
knn = KNeighborsClassifier(n_neighbors=10, 
                           weights='distance',
                           algorithm='kd_tree'
                          )
knn.fit(X_train_transformed, y_train)
Q3_7 = knn.predict(X_test_transformed)

knn = KNeighborsClassifier(n_neighbors=10, 
                           weights='distance',
                           algorithm='ball_tree'
                          )
knn.fit(X_train_transformed, y_train)
Q3_8 = knn.predict(X_test_transformed)

knn = KNeighborsClassifier(n_neighbors=10,
                           weights='distance',
                            algorithm='brute'
                          )
knn.fit(X_train_transformed, y_train)
Q3_9 = knn.predict(X_test_transformed)

knn = KNeighborsClassifier(n_neighbors=10,
                           weights='uniform',
                           algorithm='kd_tree'
                          )
knn.fit(X_train_transformed, y_train)
Q3_10 = knn.predict(X_test_transformed)

knn = KNeighborsClassifier(n_neighbors=10,
                           weights='uniform',
                           algorithm='ball_tree'
                          )
knn.fit(X_train_transformed, y_train)
Q3_11 = knn.predict(X_test_transformed)

knn = KNeighborsClassifier(n_neighbors=10,
                           weights='uniform',
                            algorithm='brute'
                          )
knn.fit(X_train_transformed, y_train)
Q3_12 = knn.predict(X_test_transformed)

print('Accuracy for k = 10, weights = distance, algorithm = kd_tree', accuracy_score(y_test, Q3_7))
print('Accuracy for k = 10, weights = distance, algorithm = ball_tree', accuracy_score(y_test, Q3_8))
print('Accuracy for k = 10, weights = distance, algorithm = brute', accuracy_score(y_test, Q3_9))
print('Accuracy for k = 10, weights = uniform, algorithm = kd_tree', accuracy_score(y_test, Q3_10))
print('Accuracy for k = 10, weights = uniform, algorithm = ball_tree', accuracy_score(y_test, Q3_11))
print('Accuracy for k = 10, weights = uniform, algorithm = brute', accuracy_score(y_test, Q3_12))

Accuracy for k = 10, weights = distance, algorithm = kd_tree 0.9346
Accuracy for k = 10, weights = distance, algorithm = ball_tree 0.9346
Accuracy for k = 10, weights = distance, algorithm = brute 0.9346
Accuracy for k = 10, weights = uniform, algorithm = kd_tree 0.9328
Accuracy for k = 10, weights = uniform, algorithm = ball_tree 0.9328
Accuracy for k = 10, weights = uniform, algorithm = brute 0.9328


In [17]:
knn = KNeighborsClassifier(n_neighbors=15, 
                           weights='distance',
                           algorithm='kd_tree'
                          )
knn.fit(X_train_transformed, y_train)
Q3_13 = knn.predict(X_test_transformed)

knn = KNeighborsClassifier(n_neighbors=15, 
                           weights='distance',
                           algorithm='ball_tree'
                          )
knn.fit(X_train_transformed, y_train)
Q3_14 = knn.predict(X_test_transformed)

knn = KNeighborsClassifier(n_neighbors=15,
                           weights='distance',
                            algorithm='brute'
                          )
knn.fit(X_train_transformed, y_train)
Q3_15 = knn.predict(X_test_transformed)

knn = KNeighborsClassifier(n_neighbors=15,
                           weights='uniform',
                           algorithm='kd_tree'
                          )
knn.fit(X_train_transformed, y_train)
Q3_16 = knn.predict(X_test_transformed)

knn = KNeighborsClassifier(n_neighbors=15,
                           weights='uniform',
                           algorithm='ball_tree'
                          )
knn.fit(X_train_transformed, y_train)
Q3_17 = knn.predict(X_test_transformed)

knn = KNeighborsClassifier(n_neighbors=15,
                           weights='uniform',
                            algorithm='brute'
                          )
knn.fit(X_train_transformed, y_train)
Q3_18 = knn.predict(X_test_transformed)

print('Accuracy for k = 15, weights = distance, algorithm = kd_tree', accuracy_score(y_test, Q3_13))
print('Accuracy for k = 15, weights = distance, algorithm = ball_tree', accuracy_score(y_test, Q3_14))
print('Accuracy for k = 15, weights = distance, algorithm = brute', accuracy_score(y_test, Q3_15))
print('Accuracy for k = 15, weights = uniform, algorithm = kd_tree', accuracy_score(y_test, Q3_16))
print('Accuracy for k = 15, weights = uniform, algorithm = ball_tree', accuracy_score(y_test, Q3_17))
print('Accuracy for k = 15, weights = uniform, algorithm = brute', accuracy_score(y_test, Q3_18))

Accuracy for k = 15, weights = distance, algorithm = kd_tree 0.9325
Accuracy for k = 15, weights = distance, algorithm = ball_tree 0.9325
Accuracy for k = 15, weights = distance, algorithm = brute 0.9325
Accuracy for k = 15, weights = uniform, algorithm = kd_tree 0.9311
Accuracy for k = 15, weights = uniform, algorithm = ball_tree 0.9311
Accuracy for k = 15, weights = uniform, algorithm = brute 0.9311


In [18]:
knn = KNeighborsClassifier(n_neighbors=20, 
                           weights='distance',
                           algorithm='kd_tree'
                          )
knn.fit(X_train_transformed, y_train)
Q3_19 = knn.predict(X_test_transformed)

knn = KNeighborsClassifier(n_neighbors=20, 
                           weights='distance',
                           algorithm='ball_tree'
                          )
knn.fit(X_train_transformed, y_train)
Q3_20 = knn.predict(X_test_transformed)

knn = KNeighborsClassifier(n_neighbors=20,
                           weights='distance',
                            algorithm='brute'
                          )
knn.fit(X_train_transformed, y_train)
Q3_21 = knn.predict(X_test_transformed)

knn = KNeighborsClassifier(n_neighbors=20,
                           weights='uniform',
                           algorithm='kd_tree'
                          )
knn.fit(X_train_transformed, y_train)
Q3_22 = knn.predict(X_test_transformed)

knn = KNeighborsClassifier(n_neighbors=20,
                           weights='uniform',
                           algorithm='ball_tree'
                          )
knn.fit(X_train_transformed, y_train)
Q3_23 = knn.predict(X_test_transformed)

knn = KNeighborsClassifier(n_neighbors=20,
                           weights='uniform',
                           algorithm='brute'
                          )
knn.fit(X_train_transformed, y_train)
Q3_24 = knn.predict(X_test_transformed)

print('Accuracy for k = 20, weights = distance, algorithm = kd_tree', accuracy_score(y_test, Q3_19))
print('Accuracy for k = 20, weights = distance, algorithm = ball_tree', accuracy_score(y_test, Q3_20))
print('Accuracy for k = 20, weights = distance, algorithm = brute', accuracy_score(y_test, Q3_21))
print('Accuracy for k = 20, weights = uniform, algorithm = kd_tree', accuracy_score(y_test, Q3_22))
print('Accuracy for k = 20, weights = uniform, algorithm = ball_tree', accuracy_score(y_test, Q3_23))
print('Accuracy for k = 20, weights = uniform, algorithm = brute', accuracy_score(y_test, Q3_24))

Accuracy for k = 20, weights = distance, algorithm = kd_tree 0.9315
Accuracy for k = 20, weights = distance, algorithm = ball_tree 0.9315
Accuracy for k = 20, weights = distance, algorithm = brute 0.9315
Accuracy for k = 20, weights = uniform, algorithm = kd_tree 0.9288
Accuracy for k = 20, weights = uniform, algorithm = ball_tree 0.9288
Accuracy for k = 20, weights = uniform, algorithm = brute 0.9288


In [19]:
knn = KNeighborsClassifier(n_neighbors=30, 
                           weights='distance',
                           algorithm='kd_tree'
                          )
knn.fit(X_train_transformed, y_train)
Q3_25 = knn.predict(X_test_transformed)

knn = KNeighborsClassifier(n_neighbors=30, 
                           weights='distance',
                           algorithm='ball_tree'
                          )
knn.fit(X_train_transformed, y_train)
Q3_26 = knn.predict(X_test_transformed)

knn = KNeighborsClassifier(n_neighbors=30,
                           weights='distance',
                            algorithm='brute'
                          )
knn.fit(X_train_transformed, y_train)
Q3_27 = knn.predict(X_test_transformed)

knn = KNeighborsClassifier(n_neighbors=20,
                           weights='uniform',
                           algorithm='kd_tree'
                          )
knn.fit(X_train_transformed, y_train)
Q3_28 = knn.predict(X_test_transformed)

knn = KNeighborsClassifier(n_neighbors=30,
                           weights='uniform',
                           algorithm='ball_tree'
                          )
knn.fit(X_train_transformed, y_train)
Q3_29 = knn.predict(X_test_transformed)

knn = KNeighborsClassifier(n_neighbors=30,
                           weights='uniform',
                           algorithm='brute'
                          )
knn.fit(X_train_transformed, y_train)
Q3_30 = knn.predict(X_test_transformed)

print('Accuracy for k = 30, weights = distance, algorithm = kd_tree', accuracy_score(y_test, Q3_25))
print('Accuracy for k = 30, weights = distance, algorithm = ball_tree', accuracy_score(y_test, Q3_26))
print('Accuracy for k = 30, weights = distance, algorithm = brute', accuracy_score(y_test, Q3_27))
print('Accuracy for k = 30, weights = uniform, algorithm = kd_tree', accuracy_score(y_test, Q3_28))
print('Accuracy for k = 30, weights = uniform, algorithm = ball_tree', accuracy_score(y_test, Q3_29))
print('Accuracy for k = 30, weights = uniform, algorithm = brute', accuracy_score(y_test, Q3_30))

Accuracy for k = 30, weights = distance, algorithm = kd_tree 0.9282
Accuracy for k = 30, weights = distance, algorithm = ball_tree 0.9282
Accuracy for k = 30, weights = distance, algorithm = brute 0.9282
Accuracy for k = 30, weights = uniform, algorithm = kd_tree 0.9288
Accuracy for k = 30, weights = uniform, algorithm = ball_tree 0.926
Accuracy for k = 30, weights = uniform, algorithm = brute 0.926


The best result I was able to find was an accuracy of 93.54% using k = 20, distance weight with kd_tree, ball_tree and brute algorithm.
<br>
The accuracy scores tend to decrease as k increases (overfitting).
<br>
The accuracy scores tend to stay the same while using different algorithms (kd_tree, ball_tree and brute).
<br>
The accuracy scores tend to increase for weights = distance than uniform.

In [20]:
# For grader use

maxScore = maxScore +12

# Insert grade here (each item out of 4)
# G[3,1] =
# G[3,2] = 
# G[3,3] =
# G[3,4] = 

In [ ]:
# For grader use

# Total Grade Calculation

rawScore = np.sum(G)
score = rawScore*100/maxScore